snapshot/git/gitfiler/cloner.go

```diff
@@ -0,0 +1,45 @@
+package gitfiler
+
+import (
+	"fmt"
+	"log"
+	"os/exec"
+
+	"github.com/scootdev/scoot/os/temp"
+	"github.com/scootdev/scoot/snapshot/git/repo"
+)
+
+// A Reference Repository is a way to clone repos locally so that the clone takes less time and disk space.
+// By passing --reference <local path> to a git clone, the clone will not copy the whole ODB but instead
+// hardlink. This means the clone is much faster and also takes very little extra hard disk space.
+// Cf. https://git-scm.com/docs/git-clone
+
+// cloner clones a repo using --reference based on a reference repo
+type refCloner struct {
+	refPool   *RepoPool
+	clonesDir *temp.TempDir
+}
+
+// Get gets a repo with git clone --reference
+func (c *refCloner) Get() (*repo.Repository, error) {
+	ref, err := c.refPool.Get()
+	defer c.refPool.Release(ref, err)
+	if err != nil {
+		return nil, err
+	}
+
+	cloneDir, err := c.clonesDir.TempDir("clone-")
+	if err != nil {
+		return nil, err
+	}
+
+	// We probably ought to use a separate git dir so that processes can't mess up .git
```

Comment:
We should use TODOs if we really want to revisit this code and add that restriction. Otherwise I know I won't remember and won't be able to grep for TODO/FIXME.